### PR TITLE
Add conflicts cache for tx_pool to record conflicted transactions

### PR DESCRIPTION
**Pull request by chenyukang**
### What problem does this PR solve?

Problem Summary:

Starting from 112.0, we don't allow conflicted transactions stay in txpool at the same time, this may have a regression impact for compact block replay, since we reject any double-spending transaction when RBF is disable or RBF check failed.

### What is changed and how it works?

We add a new pool called `conflicts_cache` into `TxPool` in this PR, we record the conflicted transaction in these scenarios:

1. RBF is enabled, record those replaced transactions into `conflicts_cache`
2. RBF is enabled, but RBF check rule failed with conflicts, record old transaction into `conflicts_cache`
3. RBF is disabled, but there is a conflict happening when `pre_check`, add the new transaction into `conflicts_cache` after verification.

What's Changed:

### Related changes

- PR to update `owner/repo`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code ci-runs-only: [ quick_checks,linters ]

Side effects

- Performance regression
- Breaking backward compatibility

### Release note 

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

File: tx-pool/src/pool.rs [hunk 1]
```diff
@@ -380,11 +388,14 @@ impl TxPool {
...
         proposal_id: &ProposalShortId,
     ) -> Option<TransactionView> {
-        self.get_tx_from_pool(proposal_id).cloned().or_else(|| {
-            self.committed_txs_hash_cache
-                .peek(proposal_id)
-                .and_then(|tx_hash| self.snapshot().get_transaction(tx_hash).map(|(tx, _)| tx))
-        })
+        self.get_tx_from_pool(proposal_id)
+            .cloned()
+            .or_else(|| self.conflicts_cache.peek(proposal_id).cloned())
```
Comment (eval-exec) [inline tx-pool/src/pool.rs:393]: When using `get_tx_with_cycles`, should it use `peek` or `get`?

> Unlike `get`, `peek` does not update the least recently used (LRU) list, so the position of the key will remain unchanged.

Comment (chenyukang) [reply tx-pool/src/pool.rs:393]: I think we should use `peek` here (the difference is very trivial),
`get_tx_from_pool_or_store` is with `&self`, if use `get` we need `mut self`, 
I guess the same reason applied to `committed_txs_hash_cache.peek(..)` .

File: sync/src/relayer/mod.rs [hunk 2]
```diff
@@ -447,8 +447,20 @@ impl Relayer {
 
         if !short_ids_set.is_empty() {
             let tx_pool = self.shared.shared().tx_pool_controller();
-
+            let short_ids_set_clone = short_ids_set.clone();
```
Comment (quake) [inline sync/src/relayer/mod.rs:450]: suggest to use `log_enabled` and avoids clone when the debug level is disabled.

File: tx-pool/src/process.rs [hunk 3]
```diff
@@ -135,7 +155,7 @@ impl TxPoolService {
                 for id in conflicts.iter() {
                     let removed = tx_pool.pool_map.remove_entry_and_descendants(id);
                     for old in removed {
-                        debug!(
+                        eprintln!(
```
Comment (quake) [inline tx-pool/src/process.rs:158]: why use eprintln here?

Comment (chenyukang) [reply tx-pool/src/process.rs:158]: it's forgotten after debug code, rollbacked.

File: sync/src/relayer/mod.rs (hunk 2 above)
Comment (chenyukang) [reply sync/src/relayer/mod.rs]: fixed.

**Review by eval-exec — DISMISSED** [review]

**Review by doitian — APPROVED** [review]

**Review by zhangsoledad — APPROVED** [review]

**Pull request merged**